### ISSUE TITLE
[Perf][B300] Evaluate a vocabulary-tiled GRPO backward

### ISSUE THREAD

**Issue by heiheiha798** [heiheiha798/flash-linear-attention#6]
## Summary

Evaluate a two-dimensional, vocabulary-tiled Triton backward for the public `fused_grpo_loss` operator on NVIDIA B300. This is an unmeasured optimization hypothesis, not a performance claim.

The current backward launches one CTA per completion token with a 32,768- or 65,536-lane live vector at the frozen public test vocabularies. The forward has already computed and saved each row's LSE; after the scalar GRPO `dlogp` is known, vocabulary-gradient tiles are independent. A `(completion_row, vocabulary_tile)` grid could expose much more parallel work and reduce per-CTA live state without changing the loss equations.

This is an operator-level RL fine-tuning improvement. The repository does not contain a GRPO trainer or model callsite, so any result must be stated for the public fused loss API and must not be extrapolated to unmeasured trainer throughput.

Baseline source: `upstream/main@e47d5d20aeb5989b58a3738b872e7c288a9fb75f`.

## Mandatory baseline and viability gate

Before any source or benchmark-fixture commit:

1. Run the complete untouched `pytest tests/modules/test_grpo.py -q` on B300. A deterministic failure or OOM ends this cycle; do not repair, skip, narrow, or edit the oracle in this performance change.
2. Profile the untouched fused GRPO forward+backward at B2/T16 and B2/T1024 for V=32,000 and 65,536. Use a full operator trace and full/source-correlated NCU captures of `grpo_fwd_kernel`, `grpo_bwd_kernel`, and the final last-token zero fill.
3. Proceed only if backward is a material part of the public operator and NCU shows enough giant-vector register/local-memory/occupancy or eligible-warp headroom to support the acceptance thresholds below. If backward is already at the DRAM roofline without a resource limit, the full-operator bound is insufficient, or the long-completion geometry has no headroom, record a no-go before source edits.

Use a small ignored harness for this baseline profiling because the operator is not yet in `benchmarks/ops/registry.py`. Keep raw reports and logs outside git.

## Bounded design

If the viability gate passes:

1. Extend the existing forward kernel with one internal fp32 `saved_logp` output of shape `[B, L]`. For active rows, store the already-computed `x_target - lse`; for masked rows, store zero so the buffer is fully initialized under NaN poisoning. Save it in the autograd context. This costs exactly `4 * B * L` bytes, one scalar per completion slot, not per vocabulary element.
2. Replace the giant backward launch for the measured NVIDIA large-vocabulary region with a grid `(B * L, ceil(V / BV))`. Select one coherent `BV`/warp configuration such as 1K, 2K, or 4K from profile evidence, not a B/T/V lookup table.
3. Each CTA reloads the saved row `logp`, LSE, reference log probability, advantage, completion mask, beta, and upstream gradient; computes the unchanged fp32 scalar `dlogp`; then reads and writes only its disjoint contiguous vocabulary slice. There are no atomics, inter-CTA reductions, or scratch tensors proportional to vocabulary size.
4. Preserve the existing final zero fill for the unused last-token logits. Every tile of a masked row must write zero, and every out-of-place gradient element must be initialized.

The saved log probability is required for `inplace=True`. Without it, a CTA containing the target column could overwrite that logit before sibling CTAs load it. A Triton `debug_barrier` is only intra-CTA and cannot solve this cross-CTA race. Once `logp` is saved, each CTA reads and overwrites only its own vocabulary slice.

Cast both program IDs and all row, physical-row, batch, vocabulary-tile, stride, and base computations to int64 before address arithmetic. Do not use block pointers, atomics, reduced precision, new fast-math behavior, global synchronization, or a vocabulary-sized copy.

Keep the existing backward as a simple fallback for coherent losing/small-vocabulary regions and for non-NVIDIA backends. The new kernel code should remain ordinary Triton and must not add or reuse an unmerged SM103 device helper. Keep the forward reduction, GRPO equations, exp helper, public signature/output, last-token semantics, and dispatched Ascend backend unchanged.

## Expected files and frozen scope

- `fla/modules/grpo.py`: internal saved-logp output/context state, tiled backward kernel, and conservative dispatch.
- `benchmarks/ops/registry.py`: one repository-native `fused_grpo_loss` registration created as a fixture-only commit before kernel edits.
- `tests/modules/test_grpo.py`: byte-for-byte frozen from upstream for the entire cycle.

No model/config, Ascend backend, old-logps GRPO path, public API, reference, tolerance, skip, seed, or numeric-flag change is in scope.

## Frozen correctness and fixture protocol

The immutable promotion oracle is the complete upstream `tests/modules/test_grpo.py`. It compares fused loss, returned KL, and full `dlogits` against the fp32 PyTorch reference for B=2; T=16/1024; V=32,000/65,536; bf16; `inplace=True/False`; partial masks; nonuniform upstream gradients; and 100 fresh-input repetitions. Module-test NaN poisoning checks full out-of-place initialization. The separate old-logps GRPO tests remain part of the full gate even though their implementation is untouched.

On B300, the existing `T=4096` parametrization is skipped by the frozen `IS_NVIDIA_HOPPER` condition. Do not edit that condition or use T=4096 as promotion correctness evidence. Longer-completion timings are exploratory only.

After baseline viability, create a registry-only fixture commit while `fla/modules/grpo.py` is still identical to the source baseline:

- Generate logits `[B, T+1, V]` bf16, reference log probabilities `[B, T]`, input IDs `[B, T+64]` int64 clamped to `[0, V)`, advantages `[B]` fp32, and completion masks `[B, T]` with deterministic active and masked regions.
- Use the public defaults `inplace=False` and `save_kl=False`, with beta 0.04, and set tuple handling for the tensor return.
- Give the op GRPO-specific default shapes and `test_file='tests/modules/test_grpo.py'`.
- Run the untouched source and full frozen test file green, then freeze this fixture commit SHA.

Because upstream does not contain this registry entry, all repository-native `verify` and `run` baseline comparisons must use the frozen fixture SHA as `--base`, not `upstream/main`. The underlying source baseline remains `e47d5d20aeb5989b58a3738b872e7c288a9fb75f`.

Supplement the immutable gate with an ignored reference harness for public-default `save_kl=False`, `completion_mask=None`, fully active masks, half masks, and ragged prefix/suffix masks. Compare loss and full gradients in both in-place and out-of-place modes where applicable. This supplemental evidence may not modify the committed test or carry a relaxed tolerance.

Per iteration, run the full unfiltered gate through:

`python -m benchmarks.ops.verify --op fused_grpo_loss --base <fixture-sha> --modes fwd fwdbwd`

Promotion also requires dependent-test discovery, the selected dependent tests, CPU compilation/static checks, and changed-file pre-commit hooks.

## Repository-native benchmark plan

Use a retained matrix with B in {2, 4, 8}, T in {16, 128, 512, 1024}, and V in {32,000, 65,536}, pruning only for memory while retaining both exactly gated B2/T16 and B2/T1024 endpoints for each vocabulary. These four exact endpoints are the promotion headline; scaling shapes cannot hide an endpoint regression.

Benchmark public-default out-of-place mode in `fwd` and `fwdbwd` through `benchmarks.ops.verify` and `benchmarks.ops.run` against the frozen fixture SHA. Use identical tensors, dtype, clocks, environment, warmup, repetitions, and fully warmed autotune state in the same B300 allocation; interleave baseline and candidate where practical.

The native runner must not repeatedly benchmark destructive in-place backward on the same logits. If secondary in-place latency is reported, use a repository-local one-call or clone-per-iteration harness and disclose clone/allocation timing semantics. In-place performance is not required for promotion, but its frozen correctness is mandatory.

GRPO has no dense/varlen sequence-layout branch. Its relevant layout axis is completion masking, covered by the supplemental fully active, half-masked, and ragged prefix/suffix distributions. Do not invent a dense/varlen model or trainer claim.

Native p20/median/p80 output is supplemental. Final evidence must report median, mean, standard deviation, minimum, p10, p90, sample count, and equal-weight geomean speedup, never best runs.

## B300 NCU evidence

Capture baseline and candidate at all four exact endpoints, filtering the old/new backward plus forward and last-token fill. Use both full metrics with PM sampling/warp states and source-correlated `SourceCounters`, one captured launch per report where applicable. Preserve raw `.ncu-rep` files, exports, exact commands, source/fixture/candidate SHAs, B300 capability, driver/CUDA/PyTorch/Triton/NCU versions, clocks/idle evidence, and Slurm job IDs outside git. Do not generate or claim checksums.

Report per-kernel and summed forward+backward duration, grid/CTA/wave count, registers per thread, achieved/theoretical occupancy and limiting factor, local loads/stores and spills, active/eligible warps, issue stalls, DRAM/L2 bytes/throughput/hit behavior, and source hot instructions. Confirm the expected number of vocabulary tiles launched and that full gradient bytes were written; an implausibly large win is a skipped-work or race suspect.

Promotion requires the profile to show that smaller live vectors and more independent CTAs improve occupancy/eligible warps or remove local spill/stall cost without scalar-reload, scheduling, spill, or memory-traffic overhead erasing the full-operator benefit.

## Acceptance and stop conditions

- The complete frozen test file passes unchanged. Both in-place and out-of-place full gradients match; masked rows and final rows are initialized; there is no tolerance, reference, seed, precision, numeric-flag, skip, or environment-suppression change.
- Across five same-session B300 comparisons, the new backward kernel is at least 1.30x faster in median at each of the four exact B2 endpoints.
- Public out-of-place forward+backward achieves at least 1.15x equal-weight geomean across those endpoints and at least 1.10x at every endpoint.
- Forward including the `4 * B * L` side-buffer allocation/store remains within 2% measured noise. No retained scaling shape regresses by more than 3% outside measured noise.
- No `O(B * L * V)` auxiliary allocation, atomics, global synchronization, or second reduction is introduced.
- NCU explains the gain through the stated resource/parallelism mechanism and confirms complete execution.

Use the legacy fallback for a coherent losing region. Record a no-go rather than broadening scope if the full baseline oracle is red, backward is not material, NCU shows a bandwidth roof without resource headroom, tiled CTAs lose at T=1024, in-place correctness requires a vocabulary-sized copy or global synchronization, gains require a narrow B/T lookup table, or any acceptance threshold is missed.

## Duplicate audit and risks

The 2026-08-21 audit found no open upstream GRPO optimization issue, PR, or matching remote branch. Merged #189 introduced the operator, #377 introduced in-place backward, #1076 fixed the separate old-logps implementation, and #927 added the separately dispatched Ascend backend. None implements vocabulary-tiled backward. Open benchmark-profile work may cause a mechanical registry conflict but does not add GRPO or change this kernel.

The principal correctness risk is the in-place cross-CTA target-logit race; forward-saved logp and the existing 100-repeat oracle bound it. Performance risks are redundant scalar loads, scheduling overhead at long T, and a tile choice that trades lower live state for more memory traffic or occupancy loss. The profile-first gate, exact endpoint thresholds, and retained legacy fallback decide those risks.
